src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp

```diff
@@ -0,0 +1,230 @@
+
+    // Optional pattern to capture alibi slopes (based on pattern from bloom)
+    auto alibi = pattern::any_input();
+    auto sdpa_mask = pattern::wrap_type<v1::Multiply>({pattern::any_input(), alibi});
+    sdpa_mask = pattern::wrap_type<v1::Reshape>({sdpa_mask, pattern::any_input()});
+    sdpa_mask = pattern::wrap_type<v1::Reshape>({sdpa_mask, pattern::any_input()});
+    sdpa_mask = pattern::wrap_type<v1::Select>({pattern::any_input(), pattern::any_input(), sdpa_mask});
+
+    auto q = pattern::any_input();
+    auto sdpa = pattern::wrap_type<v13::ScaledDotProductAttention>({
+        q,
+        std::make_shared<pattern::op::Or>(OutputVector{k_concat, k_shaped, k_shaped_transposed, k_simply_shaped}),
+        std::make_shared<pattern::op::Or>(OutputVector{v_concat, v_shaped, v_shaped_transposed, v_simply_shaped}),
+        std::make_shared<pattern::op::Or>(OutputVector{sdpa_mask, pattern::any_input()})
+    });
+
+    ov::matcher_pass_callback callback = [OV_CAPTURE_CPY_AND_THIS, &kv_parameters, &model_remaining_params, &sliding_window, &parameters_to_remove, &assignes_to_remove](ov::pass::pattern::Matcher& m) {
+        std::cout << "____" << matcher_name << "___Matched___" << std::endl;
+        OPENVINO_ASSERT(m.get_pattern_value_map().find(sdpa) != m.get_pattern_value_map().end());
+        const auto& pattern_map = m.get_pattern_map();
+        OPENVINO_ASSERT(pattern_map.find(sdpa) != pattern_map.end());
+        auto real_q = pattern_map.at(q);
+        // Why do have double check here?
+
+        // takes option that has 4D instead of fine-grained Reshape analysis
+        // it avoids complication in the pattern, but we don't really have many options
+        auto take_4d = [OV_CAPTURE_CPY_AND_THIS](std::shared_ptr<Node> option1, std::shared_ptr<Node> option2, std::shared_ptr<Node> option3) {
+            // Question: should it be get_partial_shape() of output or input ???
+            if (pattern_map.find(option1) != pattern_map.end() && pattern_map.at(option1)->input(0).get_partial_shape().rank().get_length() == 4) {
+                return pattern_map.at(option1);
+            } else if (pattern_map.at(option2)->input(0).get_partial_shape().rank().get_length() == 4) {
+                return pattern_map.at(option2);
+            } else {
+                return pattern_map.at(option3);
+            }
+        };
+
+        auto real_k = take_4d(k_current, k_current_reshaped, k_current2);
+        auto real_v = take_4d(v_current, v_current_reshaped, v_current2);
+
+        // is_cpu() check required here.
+        // kv_cache_type required here (CURRENTLY USING STUB)
+        auto kv_cache_type = element::i64;
+        auto k_parameter = std::make_shared<v0::Parameter>(kv_cache_type, PartialShape{-1, -1, -1, -1});
+        auto v_parameter = std::make_shared<v0::Parameter>(kv_cache_type, PartialShape{-1, -1, -1, -1});
```

Comment:
TODO: The python code has is_cpu() check here. How to do a similar thing in C++?

Reply:
Fixed.